test/test_torch.py

```diff
@@ -17228,6 +17228,122 @@ def test_reciprocal_complex_extremal(self, device, dtype):
+        y = torch.tensor(-0.0)
+        expected = torch.tensor([-1.0, -0.0, -0.0, -1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor(-1.0)
+        y = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([-1.0, 1.0, -1.0, 1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        # Constant
+        x = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([-1.0, -0.0, -0.0, -1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, -1.0)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([-1.0, -0.0, -0.0, -1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, -0.0)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1.0, 0.0, -0.0, 1.0], dtype=dtype)
+        expected = torch.tensor([1.0, 0.0, 0.0, 1.0], dtype=dtype)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, 0.0)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        # Normal
+        expected = torch.tensor([[1.0, 0.0, 0.0, 1.0] * 2, [-1.0, -0.0, -0.0, -1.0] * 2], dtype=dtype).reshape(-1)
+        x = torch.tensor([-1.0, -0.0, 0.0, 1.0] * 4, dtype=dtype)
+        y = torch.tensor([[0.0] * 4, [1.0] * 4, [-0.0] * 4, [-1.0] * 4], dtype=dtype).reshape(-1)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+    @dtypes(torch.int8, torch.short, torch.int, torch.long)
+    def test_copysign_integral(self, device, dtype):
+        # integer -0 doesn't exist
+        # Result is promoted to float
+        x = torch.tensor([-1, 0, 1], dtype=dtype)
+        y = torch.tensor(-1)
+        expected = torch.tensor([-1.0, -0.0, -1.0], dtype=torch.float)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+        x = torch.tensor([-1, 0, 1], dtype=dtype)
+        y = torch.tensor(1)
+        expected = torch.tensor([1.0, 0.0, 1.0], dtype=torch.float)
+        self.assertEqual(torch.copysign(torch.tensor(1.0), torch.copysign(x, y)),
+                         torch.copysign(torch.tensor(1.0), expected))
+
+    def _test_copysign_numpy(self, a, b):
+        torch_result = torch.copysign(a, b)
+        expected = torch.from_numpy(np.copysign(a.cpu().numpy(), b.cpu().numpy()))
+
+        # To handle inconsistencies in our type promotion and numpy's type promotion
```

Comment:
Restricting to only float types, is NumPy's type promotion different?

Reply:
There are several difference for type promotion between NumPy and PyTorch. For example:
- NumPy
np(np.float16, np.int16) -> np.float32
np(np.float16, np.int32/np.int64) -> np.float64
np(np.float32, np.int32/np.int64) -> np.float64
- PyTorch
torch(torch.float16, torch.int16) -> torch.float16
torch(torch.float16, torch.int32/torch.int64) -> torch.float16
torch(torch.float32, torch.int32/torch.int64) -> torch.float32

Restricting to only float types, there won't be any difference between NumPy and PyTorch.

Reply:
Sounds good.